File: app/src/main/java/com/github/flickr/dependency/DownloadModule.java
package com.github.flickr.dependency;

import android.content.Context;
import android.support.annotation.NonNull;

import com.bumptech.glide.Glide;
import com.bumptech.glide.RequestManager;
import com.github.flickr.dependency.downloader.BitmapMediaStore;
import com.github.flickr.dependency.downloader.GlideImageDownloader;
import com.github.flickr.dependency.downloader.ImageDownloader;
import com.github.flickr.dependency.downloader.MediaStore;

import javax.inject.Singleton;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;

@Module(includes = {
        DownloadModule.Bindings.class
})
public class DownloadModule {

    @Provides
    @Singleton
    RequestManager provideGlide(@NonNull Context context) {
        return Glide.with(context);
    }

    @Module
    public interface Bindings {
        @Binds
        @Singleton
        ImageDownloader provideFragmentPresenter(@NonNull GlideImageDownloader impl);

        @Binds
        @Singleton
        MediaStore provideMediaStorePresenter(@NonNull BitmapMediaStore impl);
    }

}

File: app/src/main/java/com/github/flickr/dependency/DownloadModule.kt
package com.github.flickr.dependency

import android.content.Context
import com.bumptech.glide.Glide
import com.bumptech.glide.RequestManager
import com.github.flickr.dependency.downloader.BitmapMediaStore
import com.github.flickr.dependency.downloader.GlideImageDownloader
import com.github.flickr.dependency.downloader.ImageDownloader
import com.github.flickr.dependency.downloader.MediaStore
import dagger.Binds
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module(
        includes = [DownloadModule.Bindings::class]
)
class DownloadModule {

    @Provides
    @Singleton
    fun provideGlide(context: Context): RequestManager {
        return Glide.with(context)
    }

    @Module
    interface Bindings {
        @Binds
        @Singleton
        fun provideFragmentPresenter(impl: GlideImageDownloader): ImageDownloader

        @Binds
        @Singleton
        fun provideMediaStorePresenter(impl: BitmapMediaStore): MediaStore
    }

}